src/access/acp.test.ts

```diff
@@ -0,0 +1,196 @@
+
+const defaultMockPolicies = {
+  policies: ["https://some.pod/policies#policy"],
+  memberPolicies: ["https://some.pod/policies#memberPolicy"],
+  acrPolicies: [] as string[],
+  memberAcrPolicies: [] as string[],
+};
+
+type Rule = {
+  allOf: string[];
+  anyOf: string[];
+  noneOf: string[];
+};
+
+function mockAcr(
+  accessTo: UrlString,
+  policies = defaultMockPolicies,
+  rules?: Record<string, Rule>
+): AccessControlResource {
+  let control = createThing({ name: "access-control" });
+  control = addIri(control, rdf.type, acp.AccessControl);
+  policies.policies.forEach((policyUrl) => {
+    control = addIri(control, acp.apply, policyUrl);
+  });
+  policies.memberPolicies.forEach((policyUrl) => {
+    control = addIri(control, acp.applyMembers, policyUrl);
+  });
+
+  const acrUrl = accessTo + "?ext=acr";
+  let acrThing = createThing({ url: acrUrl });
+  policies.acrPolicies.forEach((policyUrl) => {
+    acrThing = addIri(acrThing, acp.access, policyUrl);
+  });
+  policies.memberAcrPolicies.forEach((policyUrl) => {
+    acrThing = addIri(acrThing, acp.accessMembers, policyUrl);
+  });
+
+  let acr: AccessControlResource & WithServerResourceInfo = Object.assign(
+    mockSolidDatasetFrom(acrUrl),
+    {
+      accessTo: accessTo,
+    }
+  );
+  acr = setThing(acr, control);
+  acr = setThing(acr, acrThing);
+
+  if (rules !== undefined) {
+    Object.keys(rules).forEach((policyUrl) => {
+      let policyThing = createThing({ url: policyUrl });
+      rules[policyUrl].allOf.forEach((allOfUrl) => {
+        policyThing = addIri(policyThing, acp.allOf, allOfUrl);
+        let ruleThing = createThing({ url: allOfUrl });
+        ruleThing = addIri(ruleThing, rdf.type, acp.Rule);
+        acr = setThing(acr, ruleThing);
+      });
+      rules[policyUrl].anyOf.forEach((anyOfUrl) => {
+        policyThing = addIri(policyThing, acp.anyOf, anyOfUrl);
+        let ruleThing = createThing({ url: anyOfUrl });
+        ruleThing = addIri(ruleThing, rdf.type, acp.Rule);
+        acr = setThing(acr, ruleThing);
+      });
+      rules[policyUrl].noneOf.forEach((noneOfUrl) => {
+        policyThing = addIri(policyThing, acp.noneOf, noneOfUrl);
+        let ruleThing = createThing({ url: noneOfUrl });
+        ruleThing = addIri(ruleThing, rdf.type, acp.Rule);
+        acr = setThing(acr, ruleThing);
+      });
+      acr = setThing(acr, policyThing);
+    });
+  }
+
+  return acr;
+}
+
+describe("hasInaccessiblePolicies", () => {
+  it("returns false if the ACR contains no reference to either Policies or Rules", () => {
+    const plainResource = mockSolidDatasetFrom("https://some.pod/resource");
+    const resourceWithAcr = addMockAcrTo(
+      plainResource,
+      mockAcr("https://some.pod/resource", {
+        policies: [],
+        memberAcrPolicies: [],
+        acrPolicies: [],
+        memberPolicies: [],
+      })
+    );
+    expect(internal_hasInaccessiblePolicies(resourceWithAcr)).toEqual(false);
```

Comment:
Right, but does using `toEqual` have a negative impact on performance ? My understanding was basically that `toBe` compared references, and `toEqual` values (deep equality for objects), and it made more sense to me to compare values here than references, even if the result is the same I guess for raw types. Does that make sense, or is using `toBe` a better practice ?

Reply:
It is sometimes said that JavaScript is pass-by-value for primitive types, and pass-by-reference for objects. That's incorrect: JavaScript is just pass-by-value. It just happens that the value, when we think we're being passed an object, _is_ a reference.

`.toBe` does compare values; however, in the case of objects, those values are references - and they can be different references, even if both objects those references point to have similar values. So what `toEqual` does is follow those references, and then recursively loop over the properties of both objects to compare _their_ values. So yes, it can have a performance impact.

tl;dr in practice it does indeed not matter here, but since, as you say, you want to compare values, `toBe` better captures the intent. (And also Matthieu recently highlighted that Jest also has a `toStrictEqual`, which not only checks whether the properties of both objects have the same values, but also the same types. So in the cases where you _do_ want to use `toEqual`, I think you'd almost always want to use `toStringEqual` just like you'd prefer `===` over `==`.

Reply:
Ok, thanks, that clarifies it :)

Reply:
Resolved in 93d6919951a89e7ea5f5854548dd9ff18ef77df5